### DISCUSSION TITLE
Uncaught ReferenceError: __rf_placeholder__shareScope

### DISCUSSION THREAD

**Discussion by UrielJavier** [originjs/vite-plugin-federation#159] [Q&A]
Hello, again...

The host App has this error:
Uncaught ReferenceError: __rf_placeholder__shareScope is not defined at __federation__:18:3

I know that im not giving too much information about the case, but maybe some know why this could happen. 
<img width="240" alt="image" src="https://user-images.githubusercontent.com/9280771/159807474-e71a6fb3-e53a-42cd-9bc6-92631cc0bd4c.png">

**Comment by Jiannan-dev** [comment]
I would like to know which version you are using? This definition has been removed and I suggest you use the latest version

**Reply by UrielJavier** [reply]
Im using "@originjs/vite-plugin-federation": "^1.1.3", the latest...

**Reply by Jiannan-dev** [reply]
Are you using a chunk file generated by a previous version? __rf was previously used to identify the file prefix generated by vite-plugin-federation, and was changed to __federation a long time ago, so theoretically chunks generated after using the new build will not generate anything related to __rf

**Reply by Jiannan-dev** [reply]
Or if you provide me with a reproduce this git repository, maybe something strange will happen

**Reply by UrielJavier** [reply]
The problem is at __federation__:18:3, i didnt generate any file with __rf, but in the source code im seeing this: ${getModuleMarker('shareScope')} > return type ? `__rf_${type}__${value}` : `__rf_placeholder__${value}`...

**Reply by Jiannan-dev** [reply]
This placeholder will theoretically be replaced, if not there should be a bug

**Reply by UrielJavier** [reply]
In remote-development.ts is not going to be replaced! The method is called with only one param

**Reply by murageshnad** [reply]
Hello Everyone, @UrielJavier @Zhang Jiannan
we are getting this Error in Host App, we are not able to consume the remote component that is exposed.

**index-a46f3457.js:1441 Uncaught ReferenceError: __rf_placeholder__shareScope is not defined
    at wrapShareModule (index-35a5527a.js:240420:21)
    at index-35a5527a.js:240447:64**
    
    
    **Remote App is running on port 5173 and running on load balancer trefix port 9080 and Host App is running on port 5174 and running on load balancer trefix port 8080**
    
 


  **Host App vite.config content**
  
    import federation from '@originjs/vite-plugin-federation';
import react from '@vitejs/plugin-react';
import 'dotenv/config';
import { defineConfig } from 'vite';
//import { sentryVitePlugin } from '@sentry/vite-plugin';
import tsConfigPaths from 'vite-tsconfig-paths';

export default defineConfig({
  plugins: [
    tsConfigPaths(),
    react(),
    federation({
      name: 'default',
      remotes: {
        dummy: 'http://localhost:9080/assets/remoteEntry.js',
      },
      shared: ['react', 'react-dom'],
    }),
  ],
  // Put the Sentry vite plugin after all other plugins
  // sentryVitePlugin({
  //   url: process.env.SENTRY_BUILD_URL,
  //   org: process.env.SENTRY_BUILD_ORG,
  //   project: process.env.SENTRY_BUILD_PROJECT,
  //   authToken: process.env.SENTRY_BUILD_AUTHTOKEN, // Needs `project:releases` and `org:read` scopes
  //   disable: !(process.env.CI_COMMIT_TAG && process.env.CI_COMMIT_TAG.startsWith('v')), // Only upload Sourcemaps on Release-Builds
  // }),
  envDir: './env',
  server: {
    hmr: {
      clientPort: 5173,
    },
    port: 5173,
    host: '0.0.0.0',
    watch: {
      usePolling: true,
    },
  },
  preview: {
    host: '0.0.0.0',
    port: 5173,
    strictPort: true,
  },
  css: {
    modules: {
      localsConvention: 'camelCase',
    },
  },
  build: {
    modulePreload: false,
    target: 'esnext',
    minify: false,
    cssCodeSplit: false,
  },
});


**Remote App vite.config**

import federation from '@originjs/vite-plugin-federation';
import react from '@vitejs/plugin-react';
import 'dotenv/config';
import { defineConfig } from 'vite';
//import { sentryVitePlugin } from '@sentry/vite-plugin';
import tsConfigPaths from 'vite-tsconfig-paths';

export default defineConfig({
  plugins: [
    tsConfigPaths(),
    react(), // Put the Sentry vite plugin after all other plugins
    // sentryVitePlugin({
    //   url: process.env.SENTRY_BUILD_URL,
    //   org: process.env.SENTRY_BUILD_ORG,
    //   project: process.env.SENTRY_BUILD_PROJECT,
    //   authToken: process.env.SENTRY_BUILD_AUTHTOKEN, // Needs `project:releases` and `org:read` scopes
    //   disable: !(process.env.CI_COMMIT_TAG && process.env.CI_COMMIT_TAG.startsWith('v')), // Only upload Sourcemaps on Release-Builds
    // }),
    federation({
      name: 'eer-app',
      filename: 'remoteEntry.js',
      // Modules to expose
      exposes: {
        './Button': './src/Components/Button.tsx',
      },
      shared: ['react', 'react-dom'],
    }),
  ],
  envDir: './env',
  server: {
    hmr: {
      clientPort: 5174,
    },
    port: 5174,
    host: '0.0.0.0',
    watch: {
      usePolling: true,
    },
  },
  preview: {
    host: '0.0.0.0',
    port: 5174,
    strictPort: true,
  },
  css: {
    modules: {
      localsConvention: 'camelCase',
    },
  },
  build: {
    modulePreload: false,
    target: 'esnext',
    minify: false,
    cssCodeSplit: false,
  },
});

**Plz help me on this issue.
let me know if any changes required , we are completely blocked on this issue .**

**Reply by essoufy** [reply]
Hi @murageshnad 

I see you encountered the same "Uncaught ReferenceError: __rf_placeholder__shareScope is not defined" error as I am. Have you found any solution or workaround since then? I'm also stuck on this issue and any help would be greatly appreciated.

Thanks in advance!

**Comment by Jiannan-dev** [comment]
Also, there are some examples under the repository that may be helpful to you
https://github.com/originjs/vite-plugin-federation/tree/main/packages/examples

**Reply by UrielJavier** [reply]
Yes! I used them! Thnks :)

**Comment by dciesielkiewicz** [comment]
> Hello, again...
> 
> The host App has this error: Uncaught ReferenceError: __rf_placeholder__shareScope is not defined at **federation**:18:3
> 
> I know that im not giving too much information about the case, but maybe some know why this could happen. <img alt="image" width="240" src="https://user-images.githubusercontent.com/9280771/159807474-e71a6fb3-e53a-42cd-9bc6-92631cc0bd4c.png">

@UrielJavier Not sure if that can help in your specific use case, but I had exact same issue and solved it by replacing `vite-tsconfig-paths` usage with standard `resolve: { alias: {...} }`.

**Reply by UrielJavier** [reply]
i love you, thnksss

**Reply by UrielJavier** [reply]
@ruleeeer do you know why this change can fix the problem?

**Reply by Jiannan-dev** [reply]
sry, I don't know why this is the case, it's strange, maybe a replicated project is needed to verify it

**Reply by nemonemi** [reply]
Could you please give more details as to how you've resolved it, precisely?

Also, is it possible to have both the host and remote be Typescript projects? I've followed a tutorial that used only Javascript, and it worked seamlessly, however, with Typescript that's not the case.

**Reply by paulo9mv** [reply]
Did your code worked in dev mode also? Or only for build production?
For me using `resolve: alias: {...}` worked for build production only

**Comment by He110te4m** [comment]
I encountered the same problem in the latest version (1.2.2) . Is there a viable solution?

**Reply by He110te4m** [reply]
I logged into `transform` hook and found that the module ID that appears in transform is sometimes not `\0virtual:__federation__`.
Instead, it's `virtual: xxx/__federation__` , where xxx represents the project for remote APP.
I changed the if condition to `code.replace(getModuleMarker("shareScope"), scopeCode.join(","));`.

I modify the timing of `code.replace(getModuleMarker("shareScope"), scopeCode.join(","));`.

The final code changes are as follows:

```diff
- if (id === "\0virtual:__federation__") {
+ if (/\0virtual:.*__federation__/.test(id)) {
     const scopeCode = await devSharedScopeCode.call(
       this,
       parsedOptions.devShared
     );
     return code.replace(getModuleMarker("shareScope"), scopeCode.join(","));
   }
```

**Reply by christoph-schlumpf** [reply]
Hi @He110te4m , @kirill-kalinin,
Did this solve the issue? Do I have to fork vite-plugin-federation to be able to fix it for my app?

Kind regards

**Reply by christoph-schlumpf** [reply]
Hi @He110te4m , @kirill-kalinin,
I was able to fix it: https://github.com/originjs/vite-plugin-federation/discussions/159#discussioncomment-7633645

Best regards!

**Comment by mstykow** [comment]
I encountered this issue while converting the repos (host + remote) to use absolute imports. The remote was fine using the `tsConfigPaths` plugin but in the host I needed to replace the plugin with a manual conversion in vite.config.ts:
```javascript
resolve: {
    alias: [{ find: '@', replacement: __dirname }],
},
```

My tsconfig:
```
"compilerOptions": {
        "baseUrl": ".",
        "paths": {
            "@/src/*": ["src/*"],
            "@/e2e/*": ["e2e/*"]
        }
}
```

**Comment by christoph-schlumpf** [comment]
Hi @ruleeeer ,
Thank you for this great plugin!
Unfortunately, I have this issue only in production builds. In local development, the remote gets loaded in the host app. But after doing a production build I get the same issue as discussed above (Uncaught ReferenceError: __rf_placeholder__shareScope is not defined at).

My implementation is based on https://github.com/originjs/vite-plugin-federation/pull/481, so I dynamically load the remote.

I use v1.3.3 of the vite-plugin-federate but encountered the issue in v1.3.2 too.

Any help or advice?

**Reply by christoph-schlumpf** [reply]
Hi @ruleeeer,
Finally I got it working in production builds. The solution was to add a dummy "remotes:" declaration in vite.config.js:
```
plugins: [
        federation({
            name: "default",
            remotes: {
                dummy: {
                    external: "",
                    format: "var",
                },
            },
            shared: [],
        }),
    ],
```

As soon as I did this, the error is gone and the module can be loaded in production builds too (via __federation_method_setRemote and __federation_method_getRemote).

Best regards!

**Reply by binsky08** [reply]
Thank you so much for your approach @christoph-schlumpf . Saved my week!!! :heart:

It seems, the error `__rf_placeholder__shareScope is not defined` typically arises because the module federation runtime expects a share scope to be initialized. However, when no remotes are specified in the configuration or no modules are loaded during build time, the federation runtime may skip initializing the share scope, leading to the error when an actual remote module is fetched.

By adding a dummy remote with an empty external configuration and setting its format to 'var', you ensure that the federation runtime initializes the share scope even if there are no functional remotes in the configuration during the build process. This effectively bypasses the missing initialization issue.

Really nice workaround, shortly blew my mind :D

I have no idea how to do that, but it would be very nice to have a configuration option to enable exactly that. A forced initialization of the share scope even without remotes configured.

**Reply by dev-1995** [reply]
Hey Guys, can i ask what worked for you in this case? My Federation config for remote app is
` federation({
      name: "remoteApp",
      filename: "remoteEntry.js",
      exposes: {
        "./App": "./src/App.tsx",
      },
      shared: ["react", "react-dom"],
      remotes: {
        dummy: "",
      },
    }),`

**Reply by izakdvlpr** [reply]
> Hi @ruleeeer, Finally I got it working in production builds. The solution was to add a dummy "remotes:" declaration in vite.config.js:
> 
> ```
> plugins: [
>         federation({
>             name: "default",
>             remotes: {
>                 dummy: {
>                     external: "",
>                     format: "var",
>                 },
>             },
>             shared: [],
>         }),
>     ],
> ```
> 
> As soon as I did this, the error is gone and the module can be loaded in production builds too (via __federation_method_setRemote and __federation_method_getRemote).
> 
> Best regards!

Thanks <3

**Comment by paulo9mv** [comment]
For what I could investigate, `vite-tsconfig-paths` is messing paths from module federation and vite virtual imports.
It converts the following path: `__federation__` to `virtual:/Users/username/repo/__federation__`.

However, `vite-plugin-federation` uses some "fixed" paths in `if` cases, like this:

https://github.com/originjs/vite-plugin-federation/blob/084033339d07c0d09cff34ead30a14c0346c9154/packages/lib/src/index.ts#L152

And this makes the application break. I'm not sure how this should be addressed, perhaps changing this absolute comparison to `includes(":virtual") && includes("__federation__")` or a regex approach.

Also the following line
https://github.com/originjs/vite-plugin-federation/blob/084033339d07c0d09cff34ead30a14c0346c9154/packages/lib/src/dev/remote-development.ts#L403

It does `/@fs/node_modules/.vite/deps/any-shared-lib.js?v=12345` instead of `/@fs/Users/username/repo/node_modules/.vite/deps/any-shared-lib.js?v=12345`
________
I'm checking if there's any prop that we could pass to `vite-tsconfig-paths` to ignore `__federation__`. But I honestly think the best option would be to change these lines mentioned to allow a custom path

**Reply by husayt** [reply]
Hi @paulo9mv, did you manage to find the workaround? Would very appreciate any tips, this has blocked me altogether.

Thanks

**Reply by paulo9mv** [reply]
Hi @husayt , I wasn't able to achieve an working workaround. I was investigating if adding `vite-plugin-federation` to our project would be viable, but the current state of this plugin is not so promising. ):
This bug in particular made the implementation impracticable.

We decided to abandon it for now and wait for the next steps of [rolldown](https://github.com/rolldown/rolldown/discussions/153), the rust implementation of rollup that will come with module federation and be the core blunder of vite.

**Comment by essoufy** [comment]
Hello everyone,

I'm also experiencing the "Uncaught ReferenceError: __rf_placeholder__shareScope is not defined" issue and I'm wondering if there are any updates on a solution or workaround.

Error Message:

Uncaught ReferenceError: __rf_placeholder__shareScope is not defined
    at wrapShareModule (index-b9f14579.js:46:3988)
    at index-b9f14579.js:46:4531

Here are the details of my project:

* **Environment:** Nx Workspace
* **Applications:** A remote app and a host app
* **Vite:** "^4.0.1"
* **@originjs/vite-plugin-federation:** "^1.4.0"
*

Here are my `vite.config.ts` files:
`/// <reference types="vitest" />
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import viteTsConfigPaths from 'vite-tsconfig-paths';
import federation from '@originjs/vite-plugin-federation';

export default defineConfig({
  cacheDir: '../../node_modules/.vite/rebo-apps',

  server: {
    port: 4300,
    host: 'localhost',
    fs: {
      allow: ['..']
    }
  },

  preview: {
    port: 4200,
    host: 'localhost',
  },

  plugins: [
    react(),
    viteTsConfigPaths({
      root: '../../',
    }),
    federation({
      name: 'mo-host',
      remotes: {
        remoteApp: 'http://localhost:4301/assets/remoteEntry.js',
      },
      shared: ['react', 'react-dom'],
    })
  ],
  build: {
    outDir: '../../dist/apps/host',
    target: 'esnext',
    emptyOutDir: true,
    reportCompressedSize: true,
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },
});`

Remote Application:

`/// <reference types="vitest" />
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import viteTsConfigPaths from 'vite-tsconfig-paths';
import federation from '@originjs/vite-plugin-federation';
import { resolve } from 'path';

export default defineConfig({
  cacheDir: '../../node_modules/.vite/life-app',

  server: {
    port: 4201,
    host: 'localhost',
  },

  preview: {
    port: 4301,
    host: 'localhost',
  },

  plugins: [
    react(),
    viteTsConfigPaths({
      root: '../../',
    }),
    federation({
      name: 'remote_app',
      filename: 'remoteEntry.js',
      exposes: {
        './Button':  resolve(__dirname, './src/components/Button.tsx'),
      },
      shared: ['react', 'react-dom'],
    }),
  ],
  build: {
    outDir: '../../dist/apps/remote',
    reportCompressedSize: true,
    commonjsOptions: { transformMixedEsModules: true },
    modulePreload: false,
    target: 'esnext',
    minify: false,
    cssCodeSplit: false,
  },
});`
This issue is really blocking me. Is there a planned solution or a workaround that I could try?

Thank you in advance for your help!
